File: pufferlib/ocean/dogfight/test_flight_autoace.py
"""
AutoAce Integration Tests
=========================

Tests for the AutoAce intelligent adversarial opponent at stage 20.
These tests verify that AutoAce:
1. Pursues targets when in offensive position
2. Defends when target is behind
3. Fires when in gun solution
4. Manages energy appropriately

Run: python pufferlib/ocean/dogfight/test_flight.py --test autoace_pursues
     python pufferlib/ocean/dogfight/test_flight.py --render --fps 10 --test autoace_pursues
"""

import numpy as np

from test_flight_base import (
    get_render_mode, get_render_fps, RESULTS
)


def make_autoace_env():
    """Create environment configured for AutoAce (stage 20)."""
    from pufferlib.ocean.dogfight.dogfight import Dogfight

    # Create env with curriculum enabled, set to stage 20
    env = Dogfight(obs_scheme=0, curriculum_enabled=1, curriculum_randomize=0)
    env.reset()

    # Force stage 20 (AutoAce)
    env.set_curriculum_stage(20)

    return env


def test_autoace_stage_20():
    """Verify that stage 20 spawns with AutoAce configuration."""
    env = make_autoace_env()
    env.reset()

    # Check that we're at stage 20
    stage = env.get_curriculum_stage()
    passed = stage == 20

    result = "OK" if passed else f"FAIL (stage={stage})"
    print(f"autoace_stage: stage={stage} [{'OK' if passed else 'FAIL'}]")
    RESULTS['autoace_stage'] = 20 if passed else stage


def test_autoace_pursues():
    """Test that AutoAce runs without errors at stage 20.

    This is a basic smoke test that verifies the AutoAce integration
    doesn't crash and produces valid actions.
    """
    from pufferlib.ocean.dogfight.dogfight import Dogfight

    env = Dogfight(obs_scheme=0, curriculum_enabled=1, curriculum_randomize=0)
    env.reset()

    # Force stage 20
    env.set_curriculum_stage(20)

    render_mode = get_render_mode()
    fps = get_render_fps() or 50
    dt = 1.0 / fps if render_mode else 0.02

    steps_completed = 0
    steps = int(5.0 / 0.02)  # 5 seconds at 50Hz

    for _ in range(steps):
        # Player flies straight (neutral controls)
        action = np.array([0.5, 0.0, 0.0, 0.0, -1.0], dtype=np.float32)
        obs, reward, terminated, truncated, info = env.step(action)
        steps_completed += 1

        if render_mode:
            env.render()
            import time
            time.sleep(dt)

        if np.any(terminated) or np.any(truncated):
            break

    passed = steps_completed > 10  # At least ran for a bit
    result = "OK" if passed else "FAIL"
    print(f"autoace_pursues: ran {steps_completed} steps at stage 20 [{result}]")
    RESULTS['autoace_pursues'] = passed

    env.close()


def test_autoace_defends():
    """Test that AutoAce runs multiple episodes without crashing.

    This verifies the tactical decision-making FSM handles various
    engagement scenarios without errors.
    """
    from pufferlib.ocean.dogfight.dogfight import Dogfight

    env = Dogfight(obs_scheme=0, curriculum_enabled=1, curriculum_randomize=0)

    render_mode = get_render_mode()
    fps = get_render_fps() or 50
    dt = 1.0 / fps if render_mode else 0.02

    episodes_completed = 0
    total_steps = 0

    for ep in range(5):  # Run 5 episodes
        env.reset()
        env.set_curriculum_stage(20)

        steps = int(3.0 / 0.02)  # 3 seconds per episode
        for _ in range(steps):
            # Player pursues aggressively
            action = np.array([1.0, -0.1, 0.2, 0.0, -1.0], dtype=np.float32)
            obs, reward, terminated, truncated, info = env.step(action)
            total_steps += 1

            if render_mode:
                env.render()
                import time
                time.sleep(dt)

            if np.any(terminated) or np.any(truncated):
                break

        episodes_completed += 1

    passed = episodes_completed == 5 and total_steps > 100
    result = "OK" if passed else "FAIL"
    print(f"autoace_defends: {episodes_completed} episodes, {total_steps} total steps [{result}]")
    RESULTS['autoace_defends'] = passed

    env.close()


def test_autoace_fires():
    """Test that AutoAce fires when in gun solution.

    Note: This is hard to test directly without instrumenting the C code.
    We'll check if episodes end early (possible player kill) or if
    the opponent fires (would need logging).
    """
    from pufferlib.ocean.dogfight.dogfight import Dogfight

    env = Dogfight(obs_scheme=0, curriculum_enabled=1, curriculum_randomize=0)

    # Run multiple episodes, check for early terminations
    player_deaths = 0
    episodes = 10

    for ep in range(episodes):
        env.reset()
        env.set_curriculum_stage(20)

        steps = int(10.0 / 0.02)  # 10 seconds max
        for step in range(steps):
            # Player flies in circles (defensive)
            action = np.array([0.5, -0.2, 0.5, 0.0, -1.0], dtype=np.float32)
            obs, reward, terminated, truncated, info = env.step(action)

            if np.any(terminated):
                # Could be player death, kill, or other termination
                if step < steps - 1:  # Early termination
                    player_deaths += 1
                break

    # At stage 20, AutoAce should occasionally kill the player
    # But this depends heavily on the scenario
    result = "OK" if player_deaths >= 0 else "CHECK"  # Any result is OK for now
    print(f"autoace_fires: {player_deaths}/{episodes} early terminations [{result}]")
    RESULTS['autoace_fires'] = player_deaths

    env.close()


def test_autoace_energy():
    """Test that AutoAce manages energy (extends when slow).

    Setup: Put AutoAce in a low-energy situation.
    Expected: Should extend (fly away) to rebuild energy.
    """
    from pufferlib.ocean.dogfight.dogfight import Dogfight

    env = Dogfight(obs_scheme=0, curriculum_enabled=1, curriculum_randomize=0)
    env.reset()
    env.set_curriculum_stage(20)

    # This test would require more direct state manipulation
    # For now, just verify the environment runs at stage 20

    render_mode = get_render_mode()
    fps = get_render_fps() or 50
    dt = 1.0 / fps if render_mode else 0.02

    steps = int(3.0 / 0.02)
    for _ in range(steps):
        action = np.array([0.5, 0.0, 0.0, 0.0, -1.0], dtype=np.float32)
        obs, reward, terminated, truncated, info = env.step(action)

        if render_mode:
            env.render()
            import time
            time.sleep(dt)

        if np.any(terminated) or np.any(truncated):
            break

    print(f"autoace_energy: stage 20 runs successfully [OK]")
    RESULTS['autoace_energy'] = True

    env.close()


# Test registry
TESTS = {
    'autoace_stage': test_autoace_stage_20,
    'autoace_pursues': test_autoace_pursues,
    'autoace_defends': test_autoace_defends,
    'autoace_fires': test_autoace_fires,
    'autoace_energy': test_autoace_energy,
}


if __name__ == "__main__":
    # Run all AutoAce tests
    for name, test in TESTS.items():
        print(f"\n=== {name} ===")
        test()

File: ocean/dogfight/tests/test_flight_autoace.c
/*
 * test_flight_autoace.c - 1:1 port of test_flight_autoace.py.
 *
 * 5 high-level smoke tests verifying that AutoAce (curriculum stage 20)
 * runs without errors over multiple episodes and various player inputs.
 *
 * All tests are soft.
 */
#include "test_common.h"
#include "../autopilot.h"
#include "../autoace.h"

#define DEG (3.14159265f / 180.0f)
#define RAD (180.0f / 3.14159265f)

/* Spin up an env at curriculum stage AUTOACE.  Mirrors make_autoace_env(). */
static void setup_stage20_env(TestEnv* t) {
    memset(t, 0, sizeof(*t));
    t->env.num_agents = 1;
    t->env.max_steps = 3000;
    t->env.rng = 42;
    t->env.observations = t->observations;
    t->env.actions = t->actions;
    t->env.rewards = t->rewards;
    t->env.terminals = t->terminals;
    RewardConfig rcfg = test_default_rcfg();
    init(&t->env, /*obs_scheme=*/0, &rcfg, /*curriculum_enabled=*/1, 0, 0);
    t->env.stage = CURRICULUM_AUTOACE;
    c_reset(&t->env);
    t->env.stage = CURRICULUM_AUTOACE;
    autopilot_set_mode(&t->env.opponent_ap, AP_LEVEL,
                       AP_DEFAULT_THROTTLE, 0.0f, 0.0f);
}

static int test_autoace_stage_20(void) {
    TestEnv t; setup_stage20_env(&t);
    int stage = t.env.stage;
    int passed = (stage == CURRICULUM_AUTOACE);
    printf("autoace_stage:      stage=%d [%s]\n", stage, passed ? "OK" : "FAIL");
    return 0;
}

static int test_autoace_pursues(void) {
    TestEnv t; setup_stage20_env(&t);
    int steps_completed = 0;
    int steps = 250;  /* 5 seconds at 50Hz */
    for (int s = 0; s < steps; s++) {
        float a[5] = {0.5f, 0.0f, 0.0f, 0.0f, -1.0f};
        memcpy(t.env.actions, a, sizeof(a));
        c_step(&t.env);
        steps_completed++;
        if (t.env.terminals[0]) break;
    }
    int passed = steps_completed > 10;
    printf("autoace_pursues:    ran %d steps at stage 20 [%s]\n",
           steps_completed, passed ? "OK" : "FAIL");
    return 0;
}

static int test_autoace_defends(void) {
    TestEnv t; setup_stage20_env(&t);
    int episodes_completed = 0;
    int total_steps = 0;
    for (int ep = 0; ep < 5; ep++) {
        c_reset(&t.env);
        t.env.stage = CURRICULUM_AUTOACE;
        autopilot_set_mode(&t.env.opponent_ap, AP_LEVEL,
                           AP_DEFAULT_THROTTLE, 0.0f, 0.0f);
        int steps = 150;  /* 3 seconds */
        for (int s = 0; s < steps; s++) {
            float a[5] = {1.0f, -0.1f, 0.2f, 0.0f, -1.0f};
            memcpy(t.env.actions, a, sizeof(a));
            c_step(&t.env);
            total_steps++;
            if (t.env.terminals[0]) break;
        }
        episodes_completed++;
    }
    int passed = (episodes_completed == 5) && (total_steps > 100);
    printf("autoace_defends:    %d episodes, %d total steps [%s]\n",
           episodes_completed, total_steps, passed ? "OK" : "FAIL");
    return 0;
}

static int test_autoace_fires(void) {
    TestEnv t; setup_stage20_env(&t);
    int player_deaths = 0;
    int episodes = 10;
    for (int ep = 0; ep < episodes; ep++) {
        c_reset(&t.env);
        t.env.stage = CURRICULUM_AUTOACE;
        autopilot_set_mode(&t.env.opponent_ap, AP_LEVEL,
                           AP_DEFAULT_THROTTLE, 0.0f, 0.0f);
        int steps = 500;  /* 10 seconds */
        for (int s = 0; s < steps; s++) {
            float a[5] = {0.5f, -0.2f, 0.5f, 0.0f, -1.0f};
            memcpy(t.env.actions, a, sizeof(a));
            c_step(&t.env);
            if (t.env.terminals[0]) {
                if (s < steps - 1) player_deaths++;
                break;
            }
        }
    }
    /* Same as Python — soft check, any result is OK. */
    printf("autoace_fires:      %d/%d early terminations [OK]\n",
           player_deaths, episodes);
    return 0;
}

static int test_autoace_energy(void) {
    TestEnv t; setup_stage20_env(&t);
    int steps = 150;
    int ran = 0;
    for (int s = 0; s < steps; s++) {
        float a[5] = {0.5f, 0.0f, 0.0f, 0.0f, -1.0f};
        memcpy(t.env.actions, a, sizeof(a));
        c_step(&t.env);
        ran++;
        if (t.env.terminals[0]) break;
    }
    (void)ran;
    printf("autoace_energy:     stage 20 runs successfully [OK]\n");
    return 0;
}

int main(void) {
    int fails = 0;
    fails += test_autoace_stage_20();
    fails += test_autoace_pursues();
    fails += test_autoace_defends();
    fails += test_autoace_fires();
    fails += test_autoace_energy();
    printf("\n%d hard failures\n", fails);
    return fails;
}
